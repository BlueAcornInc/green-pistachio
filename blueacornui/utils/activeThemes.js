/**
 * @package     BlueAcorn/GreenPistachio
 * @version     4.0.0
 * @author      Blue Acorn iCi <code@blueacorn.com>
 * @copyright   Copyright Â© Blue Acorn iCi. All rights reserved.
 */

const themes = require(`${process.cwd()}/gulp-config`);

export default themes.filter((theme) => theme.gulp);
