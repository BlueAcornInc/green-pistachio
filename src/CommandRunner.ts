import { exec } from 'child_process';

export default class CommandRunner {
    public execute(command: string, options: Record<string, any> = {}): Promise<{ error: string | null, stdout: string }> {
        return new Promise((resolve, reject) => {
            exec(command, options, (error, stdout) => {
                if (error) {
                    return reject({ error, stdout });
                }

                return resolve({ error, stdout });
            });
        });
    }
}