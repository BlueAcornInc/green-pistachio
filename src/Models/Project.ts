import { join } from 'path';
import { SyncHook } from 'tapable';
import debug from 'debug';
import Module from "./Module";
import Theme, { ThemeData } from "./Theme";
import ConfigLoader from './Project/ConfigLoader';
import { Configuration } from 'webpack';
import WebpackConfigFactory from '../Webpack/ConfigFactory';
const logger = debug('gpc:project');

type ProjectConstructorArgs = {
    root: string;
    themes: Theme[];
    modules: Module[];
    enabledThemes?: string[];
};

type WebpackDevServerConfig = {
    proxyUrl: string;
    devServerUrl: string;
    devServerPort: number;
};

export default class Project {
    public root: string;
    private themes: Theme[];
    private enabledThemes: Theme[];
    private modules: Module[];
    private webpackDevelopmentMode: boolean = false;
    private proxyUrl: string = '';
    private devServerConfig: Partial<WebpackDevServerConfig> = {};

    public hooks = {
        // These hooks allow for modifying various gulp configurations
        gulp: {
            babelConfig: new SyncHook(["config"]),
            babelTypeScriptConfig: new SyncHook(["config"]),
            pngSpriteConfig: new SyncHook(["config"]),
            svgSpriteConfig: new SyncHook(["config"]),
            eslintConfig: new SyncHook(["config"]),
            criticalCssConfig: new SyncHook(["config"]),
        },
        webpack: {
            config: new SyncHook(["config"]),
            entryGlobs: new SyncHook(["globs"]),
            appendConfig: new SyncHook<[Configuration[], WebpackConfigFactory]>(["webpackConfigs", "configFactory"])
        },
        configure: new SyncHook(["project"])
    };

    public experiments = {
        webpack: {
            // Set to true in the the project.configure hook to enable unstable HMR functionality
            hmr: false,
            // Set to true to enable css modules
            cssModules: false,
            // Set to true to emit files to the pub directory instead of the individual theme directories
            emitFilesToPub: false,
        }
    };

    constructor(config: ProjectConstructorArgs) {
        const { root, themes, modules, enabledThemes } = config;

        this.root = root;
        this.themes = themes;
        this.modules = modules;
        this.enabledThemes = enabledThemes
            ? themes.filter(theme => enabledThemes.includes(theme.getData().path))
            : [];
    }

    public configure() {
        if (this.enabledThemes.length > 0) {
            // Enable default themes
            for (const theme of this.enabledThemes) {
                theme.setEnabled(true);
            }
        } else {
            for (const theme of this.themes) {
                if (
                    !theme
                        .getSourceDirectory()
                        .includes(
                            join(
                                this.getRootDirectory(),
                                'vendor'
                            )
                        )
                ) {
                    theme.setEnabled(true);
                    this.enabledThemes.push(theme);
                }
            }
        }

        // Enable default modules
        for (const magentoModule of this.modules) {
            if (
                !magentoModule
                    .getSourceDirectory()
                    .includes(
                        join(
                            this.getRootDirectory(),
                            'vendor'
                        )
                    )
            ) {
                magentoModule.setEnabled(true);
            }
        }

        this.hooks.configure.call(this);
    }

    public configureTheme(themeData: Partial<ThemeData>) {
        if (!themeData.path) {
            logger(`can not configure theme without defined path: ${JSON.stringify(themeData)}`);
            return;
        }

        const theme = this.themes.find(theme => theme.getData().path === themeData.path);

        if (!theme) {
            logger(`can not find theme with path = '${themeData.path}'`);
            return;
        }

        theme.configure(themeData);

        if (themeData.hasOwnProperty('enabled')) {
            theme.setEnabled(themeData.enabled);
        }
    }

    public enableModule(moduleName: string) {
        const magentoModule = this.modules.find(magentoModule => magentoModule.getName() === moduleName);

        if (!magentoModule) {
            logger(`could not find module by name: ${moduleName}`);
            return;
        }

        magentoModule.setEnabled(true);
    }

    public getThemes() {
        return this.enabledThemes.filter(magentoTheme => magentoTheme.getEnabled());
    }

    public getAllThemes() {
        return this.themes;
    }

    public getModules() {
        return this.modules.filter(magentoModule => magentoModule.getEnabled());
    }

    public getAllModules() {
        return this.modules;
    }

    public getRootDirectory() {
        return this.root;
    }

    public getThemePubDirectories(theme: Theme) {
        return theme.getLocales().map(locale => join(
            this.getStaticDirectory(),
            theme.getData().area,
            theme.getData().path,
            locale
        ));
    }

    public getWebpackOutputDirectories(theme: Theme) {
        if (this.isWebpackDevelopmentMode() || this.experiments.webpack.emitFilesToPub) {
            return theme.getLocales().map(locale => join(
                this.getRootDirectory(),
                ...(
                    this.isWebpackDevelopmentMode()
                    ? ['static']
                    : ['pub', 'static']
                ),
                theme.getData().area,
                theme.getData().path,
                locale
            ));
        }

        return [join(theme.getSourceDirectory(), 'web')];
    }

    public getStaticDirectory() {
        return join(
            this.getRootDirectory(),
            'pub',
            'static'
        );
    }

    public isWebpackDevelopmentMode() {
        return this.webpackDevelopmentMode;
    }

    public setWebpackDevelopmentMode() {
        this.webpackDevelopmentMode = true;
    }

    public setProxyUrl(url: string) {
        this.proxyUrl = url;
    }

    public getProxyUrl(): string {
        if (this.proxyUrl.length === 0) {
            if (!debug.enabled('gpc:logger')) {
                debug.enable('gpc:logger');
            }
            logger(`A project URL must be provided in your projects ${ConfigLoader.CONFIG_FILE}`);
            logger(`\n\nExample:
module.exports = project => {
project.hooks.configure.tap("Set Product Url", () => {
    project.setProxyUrl("https://example.test");
});
}
            `);

            throw new Error(`A project URL must be provided in your projects ${ConfigLoader.CONFIG_FILE}`);
        }

        return this.proxyUrl;
    }

    public setWebpackDevServerConfig(config: Partial<WebpackDevServerConfig>) {
        if (config.proxyUrl) {
            this.proxyUrl = config.proxyUrl;
        }

        this.devServerConfig = config;
    }

    public getWebpackDevServerConfig(): WebpackDevServerConfig {
        return {
            proxyUrl: this.getProxyUrl(),
            devServerUrl: 'green-pistachio.test',
            devServerPort: 8080,
            ...(this.devServerConfig || {})
        };
    }
}
