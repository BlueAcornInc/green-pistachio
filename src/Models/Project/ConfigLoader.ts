import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import glob from 'fast-glob';
import Project from "../Project";
const logger = debug('gpc:configLoader');

export default class ConfigLoader {
    public static CONFIG_FILE = 'green-pistachio.config.js';

    constructor(private project: Project) {}

    get filePath(): string {
        return join(
            this.project.getRootDirectory(),
            ConfigLoader.CONFIG_FILE
        );
    }

    public async configExists(): Promise<boolean> {
        try {
            const configExists = await fs.stat(this.filePath);

            if (configExists) {
                return true;
            }
        } catch (err) {
            logger(`Config file can't be found at ${this.filePath}`);
        }

        return false;
    }

    public async loadConfig() {
        const configExists = await this.configExists();

        if (configExists) {
            try {
                this.interopRequire()(this.project);
            } catch (err) {
                logger(`Problem occurred while running user provided configuration file: ${err}`);
            }
        }

        await this.loadModularConfigFiles();
    }

    private async loadModularConfigFiles() {
        const filePaths = [
            ...this.project.getAllModules(),
            ...this.project.getAllThemes()
        ].map(compilableObject =>
            join(
                compilableObject.getSourceDirectory(),
                ConfigLoader.CONFIG_FILE
            )
        );

        const configFiles = await glob(filePaths);

        for (const configFile of configFiles) {
            try {
                this.interopRequire(configFile)(this.project);
            } catch (err) {
                logger(`Problem occurred while running user provided configuration file: ${err}`);
            }
        }
    }

    private interopRequire(filePath = '') {
        const configFile = filePath || this.filePath;
        const configModule = require(configFile);
        return configModule && configModule.__esModule ? configModule.default : configModule;
    }
}