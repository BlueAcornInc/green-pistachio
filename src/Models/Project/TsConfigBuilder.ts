import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import Project from "../Project";
const logger = debug('gpc:typescript:configBuilder');

export default class TsConfigBuilder {
    public getFilePath(): string {
        return join(__dirname, '..', '..', '..', 'green-pistachio.tsconfig.json');
    }

    public async configFileExists(): Promise<boolean> {
        try {
            await fs.stat(this.getFilePath());

            return true;
        } catch (err) {
            return false;
        }
    }

    public async emitConfigFile(project: Project) {
        logger(`Generating Projects tsconfig.json base`);
        const include: string[] = [];
        const paths: Record<string, string[]> = {}

        for (const module of project.getModules()) {
            const moduleSrcDir = module.getSourceDirectory().replace(project.getRootDirectory(), '');

            const modulePaths: string[] = [
                join(moduleSrcDir, 'view/frontend/web/ts/*'),
                join(moduleSrcDir, 'view/base/web/ts/*'),
                join(moduleSrcDir, 'view/adminhtml/web/ts/*'),
            ];

            paths[`${module.getName()}/*`] = modulePaths.map(modulePath => modulePath.replace(/^\//, ''));

            include.push(join(moduleSrcDir, 'view/frontend/web/ts/**/*'));
            include.push(join(moduleSrcDir, 'view/base/web/ts/**/*'));
            include.push(join(moduleSrcDir, 'view/adminhtml/web/ts/**/*'));
        }

        for (const theme of project.getAllThemes()) {
            const themeSrcDir = theme.getSourceDirectory().replace(project.getRootDirectory(), '');

            include.push(join(themeSrcDir, '**/*'));
            
            // Add theme overrides to module paths
            for (const moduleName of Object.keys(paths)) {
                paths[moduleName].unshift(
                    join(themeSrcDir, moduleName.replace(/\/\*$/, ''), 'web/ts/*').replace(/^\//, '')
                );
            }
        }

        logger(`Resolved include paths to: ${JSON.stringify(include)}`);
        logger(`Resolved module paths to: ${JSON.stringify(paths)}`);

        const writePath = this.getFilePath();
        const config = this.getBaseConfig();

        config.include = include.map(includePath => `./${includePath}`);
        config.compilerOptions.paths = paths;

        try {
            logger(`Writing config file: ${writePath}`);
            await fs.writeFile(writePath, JSON.stringify(config, null, 4));
        } catch (err) {
            logger(`Problem writing config file: ${writePath}`);
            throw err;
        }

        return writePath;
    }

    private getBaseConfig() {
        return {
            "include": [
              "./app/**/*"
            ],
            "compilerOptions": {
              "baseUrl": ".",
              "paths": {},
              "strict": true,
              "esModuleInterop": true,
              "lib": [
                "dom",
                "es2015"
              ],
              "jsx": "react",
              "emitDecoratorMetadata": true,
              "experimentalDecorators": true,
              "target": "es5",
              "allowJs": true,
              "skipLibCheck": true,
              "allowSyntheticDefaultImports": true,
              "forceConsistentCasingInFileNames": true,
              "module": "esnext",
              "moduleResolution": "node",
              "resolveJsonModule": true,
              "isolatedModules": true,
              "noEmit": true
            }
          }
    }
}