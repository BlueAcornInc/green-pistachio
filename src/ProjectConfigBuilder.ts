import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import CommandRunner from "./CommandRunner";
import Module from "./Models/Module";
import Project from "./Models/Project";
import Theme from "./Models/Theme";
const logger = debug('gpc:configBuilder');

type MagentoPaths = {
    modules: Record<string, string>;
    themes: Record<string, string>;
};

export default class ProjectConfigBuilder {
    public static CONFIG_FILE = 'green-pistachio.config.js';
    private rootDirectory: string;

    constructor(
        rootDirectory = process.cwd()
    ) {
        this.rootDirectory = rootDirectory;
    }

    public async build({ includePath = 'app' }): Promise<Project> {
        const { themes, modules } = await this.getPaths();
        const moduleObjects: Module[] = [];

        for (const module of Object.keys(modules)) {
            moduleObjects.push(
                new Module({
                    sourceDirectory: modules[module],
                    name: module
                })
            );
        }

        const themeObjects = await this.processThemes(themes);

        const project = new Project({
            root: this.rootDirectory,
            themes: themeObjects,
            modules: moduleObjects,
            includePath
        });

        const projectConfigFile = join(
            this.rootDirectory,
            ProjectConfigBuilder.CONFIG_FILE
        );

        try {
            const configExists = await fs.stat(projectConfigFile);

            if (configExists) {
                logger('loading project config file');
                try {
                    require(projectConfigFile)(project);
                } catch (err) {
                    logger(`Problem while running user provided configuration file: ${err}`);
                    throw err;
                }
            }
        } catch (err) {
            logger(`No user provided config provided, this is optional, you can create one at: ${projectConfigFile}`);
        }

        project.configure();

        return project;
    }

    private async getPaths() {
        const commandRunner = new CommandRunner();
        
        const command = `
require 'vendor/autoload.php';
echo json_encode([
    'themes' => (new \\Magento\\Framework\\Component\\ComponentRegistrar)->getPaths('theme'),
    'modules' => (new \\Magento\\Framework\\Component\\ComponentRegistrar)->getPaths('module')
]);`;

        const { stdout } = await commandRunner.execute(`php -r "${command}"`);
        const result = JSON.parse(stdout) as MagentoPaths;

        return result;
    }

    private async processThemes(themePaths: Record<string, string>): Promise<Theme[]> {
        const allThemes = Object.keys(themePaths);
        const pathsWithParents = await Promise.all(
            Object.keys(themePaths).map(async (fullThemeName) => {
                const themePath = themePaths[fullThemeName];
                const [ themeArea ] = fullThemeName.split('/');

                const data = await fs.readFile(join(themePath, 'theme.xml'));
                const matchedParent = data.toString().match(/<parent>(.*?)<\/parent>/);
                let parentThemeName;

                if (matchedParent) {
                    const [, parentPath] = matchedParent;

                    parentThemeName = `${themeArea}/${parentPath}`;
                }

                return {
                    themePath,
                    parentThemeName
                }
            })
        );

        const processedThemes: Map<string | undefined, Theme> = new Map();

        while (allThemes.length) {
            const themeName = allThemes.shift();
            
            if (themeName) {
                const [ area, ...restParts ] = themeName.split('/')
                const themePath = restParts.join('/');
                const matchedParentsPath = pathsWithParents.find(path => path.themePath === themePaths[themeName]);
                const parentThemeName = matchedParentsPath && matchedParentsPath.parentThemeName;
                const parentTheme = processedThemes.get(parentThemeName);

                if (parentTheme || !parentThemeName) {
                    processedThemes.set(
                        themeName, 
                        new Theme({
                            sourceDirectory: themePaths[themeName],
                            parent: parentTheme,
                            area: area as "frontend" | "adminhtml",
                            path: themePath
                        })
                    )
                } else {
                    allThemes.push(themeName);
                }
            }
        }

        return Array.from(processedThemes.entries()).map(([, entry]) => entry);
    }
}