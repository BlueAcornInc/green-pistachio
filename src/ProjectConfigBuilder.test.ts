import CommandRunner from './CommandRunner';
import { mocked } from 'ts-jest/utils'; 
import ProjectConfigBuilder from './ProjectConfigBuilder';
import { promises as fs } from 'fs';
import { join } from 'path';
import mock from 'mock-fs';
import Project from './Models/Project';

jest.mock('./CommandRunner');
const mockedCommandRunner = mocked(CommandRunner, true);

beforeEach(() => {
    mocked(CommandRunner).mockClear();
});

afterEach(() => {
    mock.restore();
});

describe('Project Config Builder', () => {
    it('should build on empty project', async () => {
        mockedCommandRunner.prototype.execute.mockImplementation(async () => {
            return {
                error: null,
                stdout: '{"modules": {}, "themes": {}}'
            };
        });

        const projectConfigBuilder = new ProjectConfigBuilder();
        const projectConfig = await projectConfigBuilder.build({});

        expect(projectConfig).toMatchSnapshot({
            root: expect.any(String)
        });
    });

    it('should build themes', async () => {
        mock({
            "vendor/magento/theme-frontend-blank/theme.xml": `<theme xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Config/etc/theme.xsd">
    <title>Magento Blank</title>
    <media>
        <preview_image>media/preview.jpg</preview_image>
    </media>
</theme>`,
        "vendor/magento/theme-frontend-luma/theme.xml": `<theme xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Config/etc/theme.xsd">
        <title>Magento Luma</title>
        <parent>Magento/blank</parent>
        <media>
            <preview_image>media/preview.jpg</preview_image>
        </media>
        </theme>`
        });

        mockedCommandRunner.prototype.execute.mockImplementation(async () => {
            return {
                error: null,
                stdout: '{"modules": {}, "themes": {"frontend/Magento/blank": "vendor/magento/theme-frontend-blank", "frontend/Magento/luma": "vendor/magento/theme-frontend-luma"}}'
            };
        });

        const projectConfigBuilder = new ProjectConfigBuilder();
        const projectConfig = await projectConfigBuilder.build({});

        expect(projectConfig).toMatchSnapshot({
            root: expect.any(String)
        });
    });

    it('should load user provided config file', async () => {
        mockedCommandRunner.prototype.execute.mockImplementation(async () => {
            return {
                error: null,
                stdout: '{"modules": {}, "themes": {}}'
            };
        });

        const configFilePath = join(process.cwd(), ProjectConfigBuilder.CONFIG_FILE);
        await fs.writeFile(configFilePath, `module.exports = (project) => {
            project.someUserProvidedValue = 5;
        }`);

        const projectConfigBuilder = new ProjectConfigBuilder();
        const projectConfig: Project = await projectConfigBuilder.build({});

        await fs.unlink(configFilePath);

        expect((projectConfig as Project & { someUserProvidedValue: number } ).someUserProvidedValue).toBe(5);

        expect(projectConfig).toMatchSnapshot({
            root: expect.any(String)
        });
    });
});