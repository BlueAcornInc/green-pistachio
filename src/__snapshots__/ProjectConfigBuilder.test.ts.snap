// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Config Builder should build on empty project 1`] = `
Object {
  "hooks": Object {
    "configure": SyncHook {
      "_args": Array [
        "project",
      ],
      "_call": [Function],
      "_callAsync": [Function],
      "_promise": [Function],
      "_x": Array [],
      "call": [Function],
      "callAsync": [Function],
      "compile": [Function],
      "constructor": [Function],
      "interceptors": Array [],
      "name": undefined,
      "promise": [Function],
      "tap": [Function],
      "tapAsync": [Function],
      "tapPromise": [Function],
      "taps": Array [],
    },
    "gulp": Object {
      "babelConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "babelTypeScriptConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "criticalCssConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "eslintConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "pngSpriteConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "svgSpriteConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
    },
    "webpack": Object {
      "config": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "entryGlobs": SyncHook {
        "_args": Array [
          "globs",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
    },
  },
  "includePath": "app",
  "modules": Array [],
  "root": Any<String>,
  "themes": Array [],
  "webpackDevelopmentMode": false,
}
`;

exports[`Project Config Builder should build themes 1`] = `
Object {
  "experiments": Object {
    "webpack": Object {
      "hmr": false,
    },
  },
  "hooks": Object {
    "configure": SyncHook {
      "_args": Array [
        "project",
      ],
      "_call": [Function],
      "_callAsync": [Function],
      "_promise": [Function],
      "_x": Array [],
      "call": [Function],
      "callAsync": [Function],
      "compile": [Function],
      "constructor": [Function],
      "interceptors": Array [],
      "name": undefined,
      "promise": [Function],
      "tap": [Function],
      "tapAsync": [Function],
      "tapPromise": [Function],
      "taps": Array [],
    },
    "gulp": Object {
      "babelConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "babelTypeScriptConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "criticalCssConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "eslintConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "pngSpriteConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "svgSpriteConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
    },
    "webpack": Object {
      "config": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "entryGlobs": SyncHook {
        "_args": Array [
          "globs",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
    },
  },
  "includePath": "app",
  "modules": Array [],
  "proxyUrl": "",
  "root": Any<String>,
  "themes": Array [
    Theme {
      "data": Object {
        "area": "frontend",
        "criticalCss": undefined,
        "locales": undefined,
        "path": "Magento/blank",
        "stylesheets": undefined,
      },
      "parent": undefined,
      "sourceDirectory": "vendor/magento/theme-frontend-blank",
    },
    Theme {
      "data": Object {
        "area": "frontend",
        "criticalCss": undefined,
        "locales": undefined,
        "path": "Magento/luma",
        "stylesheets": undefined,
      },
      "parent": Theme {
        "data": Object {
          "area": "frontend",
          "criticalCss": undefined,
          "locales": undefined,
          "path": "Magento/blank",
          "stylesheets": undefined,
        },
        "parent": undefined,
        "sourceDirectory": "vendor/magento/theme-frontend-blank",
      },
      "sourceDirectory": "vendor/magento/theme-frontend-luma",
    },
  ],
  "webpackDevelopmentMode": false,
}
`;

exports[`Project Config Builder should load user provided config file 1`] = `
Object {
  "hooks": Object {
    "configure": SyncHook {
      "_args": Array [
        "project",
      ],
      "_call": [Function],
      "_callAsync": [Function],
      "_promise": [Function],
      "_x": Array [],
      "call": [Function],
      "callAsync": [Function],
      "compile": [Function],
      "constructor": [Function],
      "interceptors": Array [],
      "name": undefined,
      "promise": [Function],
      "tap": [Function],
      "tapAsync": [Function],
      "tapPromise": [Function],
      "taps": Array [],
    },
    "gulp": Object {
      "babelConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "babelTypeScriptConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "criticalCssConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "eslintConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "pngSpriteConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "svgSpriteConfig": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
    },
    "webpack": Object {
      "config": SyncHook {
        "_args": Array [
          "config",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
      "entryGlobs": SyncHook {
        "_args": Array [
          "globs",
        ],
        "_call": [Function],
        "_callAsync": [Function],
        "_promise": [Function],
        "_x": undefined,
        "call": [Function],
        "callAsync": [Function],
        "compile": [Function],
        "constructor": [Function],
        "interceptors": Array [],
        "name": undefined,
        "promise": [Function],
        "tap": [Function],
        "tapAsync": [Function],
        "tapPromise": [Function],
        "taps": Array [],
      },
    },
  },
  "includePath": "app",
  "modules": Array [],
  "root": Any<String>,
  "someUserProvidedValue": 5,
  "themes": Array [],
  "webpackDevelopmentMode": false,
}
`;
